data/fintype.lean

```diff
@@ -272,11 +272,31 @@ match n, hn with
 | 1 := λ ha, ⟨λ h, λ a b, let ⟨x, hx⟩ := fintype.card_eq_one_iff.1 ha.symm in
   by rw [hx a, hx b],
     λ _, ha ▸ le_refl _⟩
-| (n+2) := λ ha, ⟨λ h, by rw ← ha at h; exact absurd h dec_trivial, 
+| (n+2) := λ ha, ⟨λ h, by rw ← ha at h; exact absurd h dec_trivial,
   (λ h, fintype.card_unit ▸ fintype.card_le_of_injective (λ _, ())
     (λ _ _ _, h _ _))⟩
 end
 
+lemma fintype.injective_iff_surjective [fintype α] {f : α → α} : injective f ↔ surjective f :=
+by haveI := classical.prop_decidable; exact
+have ∀ {f : α → α}, injective f → surjective f,
+from λ f hinj x,
+  have h₁ : image f univ = univ := eq_of_subset_of_card_le (subset_univ _)
+    ((card_image_of_injective univ hinj).symm ▸ le_refl _),
+  have h₂ : x ∈ image f univ := h₁.symm ▸ mem_univ _,
+  exists_of_bex (mem_image.1 h₂),
+⟨this, λ hsurj, injective_of_has_left_inverse ⟨surj_inv hsurj,
+    left_inverse_of_surjective_of_right_inverse (this (injective_surj_inv hsurj))
+      (right_inverse_surj_inv hsurj)⟩⟩
+
+lemma fintype.injective_iff_surjective_of_equiv [fintype α] {f : α → β} (e : α ≃ β) :
+  injective f ↔ surjective f :=
+have injective (e.symm ∘ f) ↔ surjective (e.symm ∘ f), from fintype.injective_iff_surjective,
+⟨λ hinj, by have := surjective_comp e.bijective.2 (this.1 (injective_comp e.symm.bijective.1 hinj));
+  simpa [function.comp] using this,
```

Comment:
`simpa using this` can be shortened to just `simpa`; `have := term; simpa` can be rewritten as `simpa using term`.

Reply:
changed